### ISSUE TITLE
[typedef] False positive errors on deep destructuring

### ISSUE THREAD

**Issue by gultyaev** [typescript-eslint/typescript-eslint#4725]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/typedef": [
      "warn",
          {
            "arrayDestructuring": true,
            "arrowParameter": true,
            "memberVariableDeclaration": true,
            "objectDestructuring": true,
            "parameter": true,
            "propertyDeclaration": true,
            "variableDeclaration": true,
            "variableDeclarationIgnoreFunction": true
          }
    ]
  }
}
```

[Playground](https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&code=C4TwDgpgBAIghsOAmKBeKBtAUFXUDeUAZgJYBOAzsAHJwC2EAXFFWSQHYDmUAvgDQ48hADZwqtBs1YduPLAF0sWAMaiKFKAFUKEMgUG5SlGvSYtgbLgG4DUNScnnLnG7c4Rg23fEQAKAJTMAG4A9iQAJvp40ZjAABYkFAB0RuKmfFDxiUn2EhDyaFDuwD5wAVAA9BVQADwAtHVQAMIhdGCiHBpwAEYhAK7AmeDQcOzsIYjAJCHsGQDuCcLQJMAA5BokbSHqJN1LmSFQYGQhQRHLg3G6ELZybh5eZKVIAcFhkfi20cozVJj4qQcEH4BFyph48mYz0KxWeASslWq9UaLTaHXYXV6AyGkCgo3Gk2m7CUMTwWWSgLyhUpphspNw5JyYiBhTBDDpeDu0WKj1KAGZfAB9DJsszSLiBKChCIERG1BrNVrtOCdPFYwYAIkFGpxIzGEwQRNuWDuRD67GUUxmRQ8pXKn2iZA8fTI7Ewq0BUHYplWGVW9i9PsUpvNlqJNpKCGQr1gUZQDrwTuALrd2HpBBpjg95D+3oYq34XyEouY-uZgfzcmiwawQA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9NPACZwBmIAXANogDuAhigSADShsotQARRBBQBXJBFHoCAczojRcTiF4B7JgAU2LALZwIcFPLFKQenQCNDANTZoWF3HCFJc2iGlUdaC06osAVnASQggKElKEcj4myjDaegZGMYpxKKrwKNAubnweXsapIABudg5OOe6e3r7KpejlzsG5LPkEAJIyBKoocABiogQS1YVwAL4AumNAA&tsConfig=N4XyA)

**Expected Result**

Should not print any warnings

**Actual Result**

Shows a warning (because of that I cannot switch the rule to "error" instead of "warn")

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `^5.14.0` |
| `@typescript-eslint/parser`        | `^5.14.0` |
| `TypeScript`                       | `~4.5.5` |
| `ESLint`                           | `^8.11.0` |
| `node`                             | `v16.14.0` |

**Comment by bradzacher** [comment]
Similar to https://github.com/typescript-eslint/typescript-eslint/issues/4505

**Comment by rooby** [comment]
Looking at the pull request for #4505 (https://github.com/typescript-eslint/typescript-eslint/pull/4548) it looks like it was only a fix for objects, however it seems that deep destructuring for arrays is also broken.

For example:
```
      const [[key, value]]: [string, string][] = Object.entries(someObject);
```
Gives an error for expected type annotation on the inner array.

**Comment by kmin-jeong** [comment]
@rooby @gultyaev @bradzacher 
I want to solve the issue, so I'm also watching #4505 and palyground with issue. 
I understand not only fix for objects but also error for expected type annotation should be considered.
Is there anything else to consider?
